helsingborg-stad/app-mitt-helsingborg
# Feat/coapplicant pincode ux

## Explain the changes you’ve made

Replaced symmetric key exchange with pin code generation/input.

## Explain why these changes are made

To improve stability in encryption between two users. The old symmetric system was prone to errors due to too many edge cases and difficult scenarios.

## Explain your solution

* CaseContext/Actions/Reducer etc. has been updated to use CaseEncryptionService, and check for and generate pins if needed.
* CaseOverview has been updated to handle cases with pins and use PinInputModal for inputting pins.
* Cases are now always displayed in CaseOverview even if they can't be decrypted (however button actions are disabled then).
* Code related to polling has been (mostly) purged.
* Removed DEPRECATED encryption files.

## How to test

Concrete example:

1. Checkout this branch
2. Verify that solo applicant still works as usual.
3. Verify that main applicant can start applications without waiting for any "confirmation"
4. Verify that co-applicant can follow the case progress but not act on it (no button)
5. Verify that after main applicant has signed, co applicant can input pin code. Pin code is displayed on (main applicant) case overview.
6. Verify that co-applicant, after inputting pin, can delete data/re-install app and still input pin and continue as normal.
7. Verify that the case is submitted correctly after co-applicant has signed.

## Tested environments

- [] Storybook on a iOS device/simulator.
- [] Building the Application on a Android device/simulator.

## Screenshots

Main applicant:
<img width="346" alt="image" src="https://user-images.githubusercontent.com/2890987/157879204-90804c56-4275-4ddf-87cc-4c467aac5ce4.png">

Co-applicant:
<img width="348" alt="image" src="https://user-images.githubusercontent.com/2890987/157879321-5af459c8-71bb-4e35-afc2-4e2fb9ad204a.png">

<img width="352" alt="image" src="https://user-images.githubusercontent.com/2890987/157879380-38c91d66-4246-435e-b0a5-891d2181753b.png">

<img width="353" alt="image" src="https://user-images.githubusercontent.com/2890987/157879435-254d5c43-3819-413f-8ca4-d09e623ef7f9.png">

<img width="350" alt="image" src="https://user-images.githubusercontent.com/2890987/157879475-d2a16383-717b-4325-9be9-653b293c4eb6.png">

---

## Checklist for reviewers

Here are a few points that are useful for reviewers to check.

The short list of the (maybe) most important points are:

- Does this PR work as expected?
- Does the PR properly only concern a single feature/bug/thing?
- Is the code clear and concise?
- Will the changes impact existing functionality?
- Are there any linting errors (ESLint) that should be resolved?
- Do the automatic GitHub checks pass?

However it is good if a reviewer can be as thorough as possible,
within reasonable limits of course.

### Functionality

- Does this PR work as expected?
- If this is a new feature, does it add value, or is it a sign of feature-creep?
- Will the changes impact existing functionality?
- Does the changes create inconsistencies somewhere else?
- Does the code work on different platforms if applicable?
- How does the code handle if unexpected values are passed to functions etc.?

### Readability, syntax, formatting, style

- Is the code clear and concise?
- Has styling rules (Prettier) been applied to included files?
- Are functions/variables/etc. aptly named?

### Design

- Is the code properly planned and designed?
- Is code organized/separated in a logical fashion?
- Does the code follow clean code practices to a reasonable degree?
- Is the code self-documenting enough?
- Is the code in sync with existing patterns/technologies?
- Does the code reuse existing functionality where applicable?
- Is the code designed with reusability in mind?
- Does the code have hard-coded values that can be refactored?

### Patterns, idioms, tests, best practices

- Are there any linting errors (ESLint) that should be resolved?
- Is the code properly typed (TypeScript)?
- Does the code utilize language features to their full potential?

### Debugging, testing, configuring

- Is appropriate logging used?
- Have tests been added/modified to accommodate the changes?
- Is the feature sufficiently configurable in case tweaks are required in the future?

### Performance, reliability, scalability

- Is the code well optimized?
- Does the code scale?
- Does the code handle failure scenarios?

### Security

- Is there any obvious potential for exploits and security threats?
- Does code check proper authorization where applicable?

### Etiquettes

- Are commits atomic (does only one thing per commit)?
- Does the PR properly only concern a single feature/bug/thing?
- Are commit messages well-written?
- Do the automatic GitHub checks pass?
